aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+      if (grad_output.defined()) {
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimh, outputHeight);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimw, outputWidth);
+      }
+    } break;
+    case 3: {
+      int64_t inputDepth = input.size(dimd);
+      int64_t inputHeight = input.size(dimh);
+      int64_t inputWidth = input.size(dimw);
+      int64_t outputDepth = OUTPUTSIZE(0, dim);
+      int64_t outputHeight = OUTPUTSIZE(1, dim);
+      int64_t outputWidth = OUTPUTSIZE(2, dim);
+      TORCH_CHECK(
+          outputWidth >= 0 && outputHeight >= 0 && outputDepth >= 0,
+          "Given input size per channel: (",
+          inputDepth,
+          " x ",
+          inputHeight,
+          " x ",
+          inputWidth,
+          "). "
+          "Calculated output size per channel: (",
+          outputDepth,
+          " x ",
+          outputHeight,
+          " x ",
+          outputWidth,
+          "). Output size is too small");
+      if (grad_output.defined()) {
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimd, outputDepth);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimh, outputHeight);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimw, outputWidth);
+      }
+    } break;
+    default:
+      TORCH_CHECK(false, "unexpected dim in conv_dilated shape check", dim);
+  } // switch
+
+  TORCH_CHECK_DIM_SIZE(input, ndim, dimf, weight.size(1));
+
+  if (grad_output.defined()) {
+    TORCH_CHECK(
+        grad_output.is_contiguous(), "grad_output ought to be contiguous");
+    TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimf, weight.size(0));
+  }
+
+  if (grad_input.defined() || grad_weight.defined() || grad_bias.defined()) {
+    TORCH_CHECK(
+        grad_output.defined(), "grad_output must be defined for gradients");
+  }
+
+  if (columns.defined()) {
+    TORCH_CHECK(columns.is_contiguous(), "columns needs to be contiguous");
+  }
+
+} // conv_dilated_shape_check
+
+/*
+  conv_dilated2d_all_cpu_template
+
+  Main worker. Computes tensors output, grad_input, grad_weight,
+  and/or grad_bias if non-empty.
+ */
+void conv_dilated2d_all_cpu_template(
+    Tensor& output,
+    const Tensor& input_,
+    const Tensor& weight,
+    const Tensor& bias,
+    const Tensor& grad_output_,
+    Tensor& grad_input,
+    Tensor& grad_weight,
+    Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  Tensor input = input_;
+  Tensor grad_output = grad_output_;
+
+  // Preliminary shape checks
+  TORCH_CHECK(input.defined(), "input must be defined");
+  TORCH_CHECK(weight.defined(), "weight must be defined");
+  TORCH_CHECK(
+      weight.dim() == 4,
+      "weight must be 4D tensor but got one with shape ",
+      weight.sizes());
+  auto options = input.options();
+
+  // Temporary buffers:
+  Tensor columns = at::empty({0}, options);
+
+  bool is_batch = input.dim() == 3;
+  if (is_batch) {
+    INSERT_BATCH_DIMENSION(input, 1);
```

Comment:
This is not allowed. Convolution is a read only operation on the input. You are resizing **in-place** when you do this, which means that you are mutating the input. If there is another thread accessing the input, this is very bad juju. **THIS IS A LAND BLOCKER.**

Reply:
The original code did inplace resizing as well and resizing is undone at the end of the template function.
If temporary inplace resizing is not allowed anymore, the template function can be easily changed to handle the batch and non-batch cases without resizing but at the cost of some additional checks.

Reply:
The solution seems to be in using `input.unsqueeze(0)` instead of resizing.

Reply:
You're right. In that case we don't have to land block on this, though it would be good to get it fixed.

Reply:
Fixed. The input tensors sizes are not changed anymore.